### PR TITLE
The batch bug fix of 4.6.16 to KYLIN5

### PR DESCRIPTION
**Pull request by wuzhim**

### CODE REVIEW

File: src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/rule/ProjectEliminationRule.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+@Slf4j
+public class ProjectEliminationRule extends LockEliminationRule {
+    @Override
+    protected void doHandle(LockInfo lockInfo) {
+        if (lockInfo.getProjectLocks().isEmpty()) {
+            return;
+        }
+        lockInfo.getModuleLocks().clear();
+        lockInfo.getPathLocks().clear();
+        log.info("the project locks is not empty, clean up the module locks and path locks");
```
Comment (weiwei121723) [inline src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/rule/ProjectEliminationRule.java:34]: it should log project info here

File: src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/MemoryLockUtils.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        if (isNoNeedToLock(false, store)) {
+            return;
+        }
+        List<TransactionLock> moduleLocks = getModuleLocks(project, module);
+        assert moduleLocks.size() == 1;
+        batchLock(moduleLocks.get(0).transactionUnit(), moduleLocks, false);
+    }
+
+    public static void lockAndRecord(String resPath, List<String> specifiedPath, boolean readonly){
+        // Fixme https://olapio.atlassian.net/browse/KE-41639
```
Comment (weiwei121723) [inline src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/MemoryLockUtils.java]: remove KE-XX

File: src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/LockManager.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    ProjectLock removeProjectLock(@NotNull String projectName);
+
+    ModuleLock getModuleLock(@NotNull String projectName, @NotNull ModuleLockEnum moduleName);
+
+    /** Since metadata is in json format, when we want to lock a path, we need lock multiple resources;
+     For example, if the pathName belongs to a model, we will lock model name、cc name、cc express. **/
+    List<PathLock> getPathLock(@NotNull String pathName);
+
+    String getProjectByPath(@NotNull String pathName);
```
Comment (weiwei121723) [inline src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/LockManager.java:37]: unuse interface method，should be an  util method

File: src/core-common/src/main/java/org/apache/kylin/common/persistence/transaction/TransactionOptimisticLockManager.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kylin.common.persistence.transaction;
+
+import org.apache.kylin.common.persistence.lock.TransactionLock;
+
+public class TransactionOptimisticLockManager extends TransactionPessimisticLockManager {
```
Comment (weiwei121723) [inline src/core-common/src/main/java/org/apache/kylin/common/persistence/transaction/TransactionOptimisticLockManager.java:22]: maybe call TransactionDefauleLockManager, what's the meaning of 'Optimistic'

File: src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/rule/ProjectEliminationRule.java (hunk 1 above)
Comment (wuzhim) [reply src/core-common/src/main/java/org/apache/kylin/common/persistence/lock/rule/ProjectEliminationRule.java:34]: this is just a batch pr, we will resolve this in the next batch for avoid conflict

File: src/common-service/src/main/java/org/apache/kylin/rest/cache/CacheConstant.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kylin.rest.cache;
+
+public class CacheConstant {
+    public static final String NX = "NX";
+    public static final String XX = "XX";
+    public static final String PREFIX = "Kyligence-";
```
Comment (weiwei121723) [inline src/common-service/src/main/java/org/apache/kylin/rest/cache/CacheConstant.java]: Kylin-?

**Review by eventd — APPROVED** [review]

**Comment by wuzhim** [comment]
https://opencicd.kyligence.com/blue/organizations/jenkins/OpenSource%2FKylin-CI/detail/Kylin-CI/95/pipeline/

**Pull request merged**